exchanges/okx/okx_wrapper.go

```diff
@@ -0,0 +1,1448 @@
+		URL:                  okxAPIWebsocketPublicURL,
+		ResponseCheckTimeout: exch.WebsocketResponseCheckTimeout,
+		ResponseMaxLimit:     okxWebsocketResponseMaxLimit,
+		RateLimit:            500,
+	})
+	if err != nil {
+		return err
+	}
+	return ok.Websocket.SetupNewConnection(stream.ConnectionSetup{
+		URL:                  okxAPIWebsocketPrivateURL,
+		ResponseCheckTimeout: exch.WebsocketResponseCheckTimeout,
+		ResponseMaxLimit:     okxWebsocketResponseMaxLimit,
+		Authenticated:        true,
+		RateLimit:            500,
+	})
+}
+
+// Start starts the Okx go routine
+func (ok *Okx) Start(wg *sync.WaitGroup) error {
+	if wg == nil {
+		return fmt.Errorf("%T %w", wg, common.ErrNilPointer)
+	}
+	wg.Add(1)
+	go func() {
+		ok.Run()
+		wg.Done()
+	}()
+	return nil
+}
+
+// Run implements the Okx wrapper
+func (ok *Okx) Run() {
+	if ok.Verbose {
+		log.Debugf(log.ExchangeSys,
+			"%s Websocket: %s.",
+			ok.Name,
+			common.IsEnabled(ok.Websocket.IsEnabled()))
+		ok.PrintEnabledPairs()
+	}
+
+	if !ok.GetEnabledFeatures().AutoPairUpdates {
+		return
+	}
+	err := ok.UpdateTradablePairs(context.TODO(), false)
+	if err != nil {
+		log.Errorf(log.ExchangeSys,
+			"%s failed to update tradable pairs. Err: %s",
+			ok.Name,
+			err)
+	}
+}
+
+// GetServerTime returns the current exchange server time.
+func (ok *Okx) GetServerTime(ctx context.Context, ai asset.Item) (time.Time, error) {
+	return ok.GetSystemTime(ctx)
+}
+
+// FetchTradablePairs returns a list of the exchanges tradable pairs
+func (ok *Okx) FetchTradablePairs(ctx context.Context, a asset.Item) ([]string, error) {
+	if !ok.SupportsAsset(a) {
+		return nil, fmt.Errorf("asset type of %s is not supported by %s", a, ok.Name)
+	}
+	format, err := ok.GetPairFormat(a, false)
+	if err != nil {
+		return nil, err
+	}
+	insts := []Instrument{}
+	switch a {
+	case asset.Spot:
+		insts, err = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeSpot,
+		})
+	case asset.Futures:
+		insts, err = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeFutures,
+		})
+	case asset.PerpetualSwap:
+		insts, err = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeSwap,
+		})
+	case asset.Option:
+		var underlyings []string
+		underlyings, err = ok.GetPublicUnderlyings(context.Background(), okxInstTypeOption)
+		if err != nil {
+			return nil, err
+		}
+		var instruments []Instrument
+		for x := range underlyings {
+			instruments, err = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+				InstrumentType: okxInstTypeOption,
+				Underlying:     underlyings[x],
+			})
+			if err != nil {
+				return nil, err
+			}
+			insts = append(insts, instruments...)
+		}
+	case asset.Margin:
+		insts, err = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeMargin,
+		})
+	}
+	if err != nil {
+		return nil, err
+	}
+	if len(insts) == 0 {
+		return nil, errNoInstrumentFound
+	}
+	pairsMap := map[string]bool{}
```

Comment:
Don't you think different underlying have similar instruments? Your suggestion doesn't consider such a case. I added the filter below in case instruments come as duplicates.

```
pairsMap := map[string]bool{}
	selectedPairs := []string{}
	for x := range insts {
		var pair string
		c, err := currency.NewPairFromString(insts[x].InstrumentID)
		if err != nil {
			return nil, err
		}
		pair = format.Format(c)
		if !pairsMap[pair] {
			pairsMap[pair] = true
			selectedPairs = append(selectedPairs, pair)
		}
	}
	return selectedPairs, nil
```

Reply:
If you think that's not relevant I can remove it.

Reply:
Just tested it and a collision never occurs so it can be safely removed (instrument ID should always be unique, whereas underlying ID can be the same) but since we're iteraring each asset type per method call, this area can be optimised as per above

Reply:
Okay, I will do so.